src/compiler/scala/tools/nsc/typechecker/Namers.scala

```diff
@@ -1140,6 +1140,7 @@ trait Namers extends MethodSynthesis {
       tree.tpt.defineType {
         val inferOverridden = currentRun.isScala3 &&
           !pt.isWildcard && pt != NoType && !pt.isErroneous &&
+          !(tree.isInstanceOf[ValDef] && tree.symbol.isFinal && isConstantType(rhsTpe)) &&
```

Comment:
Does it matter that it's a val?

Reply:
`final val` without explicit type is a secret handshake to not widen, that doesn't apply to `def` (https://scala-lang.org/files/archive/spec/2.13/04-basic-declarations-and-definitions.html#value-declarations-and-definitions)